### PR TITLE
[JENKINS-71326] - Support colored badges in breadcrumb dropdown

### PR DESCRIPTION
**Pull request by eduardojandre**
See [JENKINS-71326](https://issues.jenkins.io/browse/JENKINS-71326).

To solve the problem I had to:
1-Update the backend to expose the badge severity field to the API used by dropdowns.
2-Update the dropdown and side menu to use the badge severity to set the corresponding CSS class.
3-Update the CSS files so that the badge in the dropdown and side menu behaves similarly to the normal menu.

### CODE REVIEW

**Comment by mawinter69** [comment]
Can you provide a screenshot please how it looks.
Ideally it looks the same as the badges on the manage page.
image

**Comment by mawinter69** [comment]
Replaces #8798

**Comment by eduardojandre** [comment]
Sure.
sidemenu
Dropdown

**Comment by mawinter69** [comment]
The badges on the manage page have a slightly different styling.
image
image
image
image
See https://github.com/jenkinsci/jenkins/blob/2923a7785ea735b0b90a8cfb2cbf67bbf864356e/war/src/main/scss/components/_section.scss#L101 for the styling

Just an idea, we have now 3 files (5 calls) where we render a badge. Maybe we can move that into an own jelly tag e.g. `core/src/main/resources/lib/layout/badge.jelly` so we just call it via `<l:badge badge="${attrs.badge"/>`.
This gets its own styling (under `war/src/main/scss/components/_badge.scss`) which ensures that wherever it is used it looks the same. And when we adjust we have just a single place where we need to do changes. This also allows plugins to easily reuse the badge jelly and directly get it properly styled.

**Comment by eduardojandre** [comment]
Yes, I did not notice the gradient on my tests, the animation I didn't replicate because I was not sure if this was intentional or not.

But this was something that bothered me a little, having the same component replicated in multiple files, although it was previous behavior. It seemed like a bigger change and I didn't know the best way to do it.

I will try your suggestion. 

Thanks.

**Comment by eduardojandre** [comment]
I pushed another commit with changes, can you please take a look?

Now there is a jelly that is used on Manage screen and on the side menu, but the dropdown is loaded directly on javascript through ajax so I was not able to use the same jelly view on that one. (but I used the same CSS class on it)

File: war/src/main/scss/components/_badges.scss
Comment (mawinter69) [inline war/src/main/scss/components/_badges.scss:35]: this is a bit too generic. Please see https://weekly.ci.jenkins.io/design-library/Conventions/ for conventions how to name classes.
This means it should be something like `.jenkins-badge`

File: war/src/main/scss/components/_dropdowns.scss
Comment (mawinter69) [inline war/src/main/scss/components/_dropdowns.scss:203]: This is not the right way to do it.
This should be `&__badge` (see comment for templates.js)

File: war/src/main/scss/components/_section.scss
Comment (mawinter69) [inline war/src/main/scss/components/_section.scss:93]: This should be `&__badge` and be at original position in line 101

File: war/src/main/scss/components/_side-panel-tasks.scss
Comment (mawinter69) [inline war/src/main/scss/components/_side-panel-tasks.scss:93]: the name shouldn't change here

File: core/src/main/resources/lib/layout/task.jelly
Comment (mawinter69) [inline core/src/main/resources/lib/layout/task.jelly:180]: with my proposals this becomes
```
  <l:badge badge="${attrs.badge}" class="task-icon-badge"/>
```

File: core/src/main/resources/lib/layout/badge.jelly
Comment (mawinter69) [inline core/src/main/resources/lib/layout/badge.jelly:33]: use jenkins-badge

Comment (mawinter69) [inline core/src/main/resources/lib/layout/badge.jelly:33]: I would add another optional attribute to the jelly named class, that allows to pass additional classes to the span
`<span class="badge alert-${attrs.badge.severity} ${attrs.class?:''}" tooltip="${badge.tooltip}">

File: core/src/main/resources/hudson/model/ManageJenkinsAction/index.jelly
Comment (mawinter69) [inline core/src/main/resources/hudson/model/ManageJenkinsAction/index.jelly:94]: With my proposals this becomes
`<l:badge badge="${m.badge}" class="jenkins-section__item__icon__badge"/>`

File: war/src/main/js/components/dropdowns/templates.js
Comment (mawinter69) [inline war/src/main/js/components/dropdowns/templates.js:63]: with my proposals this should be
`<span class="jenkins-dropdown__item__badge jenkins-badge alert-${badgeSeverity}" tooltip="${badgeTooltip}">${badgeText}</span>

**Review by mawinter69 — CHANGES_REQUESTED** [review]

**Review by mawinter69 — APPROVED** [review]
tested locally and it looks good

File: core/src/main/resources/lib/layout/badge.jelly
Comment (NotMyFault) [inline core/src/main/resources/lib/layout/badge.jelly:4]: The year is incorrect and the company name too, unless you work for CloudBees.

File: core/src/main/resources/lib/layout/badge.jelly
Comment (NotMyFault) [inline core/src/main/resources/lib/layout/badge.jelly:25]:
```suggestion
<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler">
```
This namespace is unused.

File: core/src/main/resources/lib/layout/badge.jelly
Comment (NotMyFault) [inline core/src/main/resources/lib/layout/badge.jelly:27]:
```suggestion
    If a Badge is informed it will be displayed, otherwise nothing is done.
    
    @since TODO
```

File: core/src/main/resources/lib/layout/badge.jelly
Comment (NotMyFault) [inline core/src/main/resources/lib/layout/badge.jelly:29]:
```suggestion
      Badge to be displayed, accepts null value.
      
      @since TODO
```

File: core/src/main/resources/lib/layout/badge.jelly
Comment (NotMyFault) [inline core/src/main/resources/lib/layout/badge.jelly:32]:
```suggestion
      Additional class to be applied.
      
      @since TODO
```

**Review by NotMyFault — COMMENTED** [review]
You forgot to run `cd war/ ; yarn run lint:fix`.

**Comment by eduardojandre** [comment]
> You forgot to run `cd war/ ; yarn run lint:fix`.

Sorry, just fixed.

I applied the suggestion of "@since TODO", when the TODO will be replace by the version? Is it in the pipeline?

**Comment by NotMyFault** [comment]
> > You forgot to run `cd war/ ; yarn run lint:fix`.
> 
> Sorry, just fixed.
> 
> I applied the suggestion of "@SInCE TODO", when the TODO will be replace by the version? Is it in the pipeline?

That happens when a maintainer runs the https://github.com/jenkinsci/jenkins/blob/master/update-since-todo.py

**Review by Kevin-CB — APPROVED** [review]
Looks fine security wise

**Review by NotMyFault — APPROVED** [review]
Thanks, looks nice :)

**Review by mawinter69 — APPROVED** [review]

**Review by timja — APPROVED** [review]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
